File: solutions/swift/bob/3/Sources/Bob/Bob.swift
import Foundation

private func isShouting(_ message: String) -> Bool {
  message.contains(where: \Character.isUppercase)
    && !message.contains(where: \Character.isLowercase)
}

class Bob {
  static func response(_ message: String) -> String {
    let message = message.trimmingCharacters(in: .whitespacesAndNewlines)
    if message.isEmpty {
      return "Fine. Be that way!"
    } else if message.last == "?" {
      return isShouting(message) ? "Calm down, I know what I'm doing!" : "Sure."
    } else {
      return isShouting(message) ? "Whoa, chill out!" : "Whatever."
    }
  }
}

File: solutions/haskell/bob/3/src/Bob.hs
module Bob (responseFor) where

import qualified Data.Char as C
import Data.List

responseFor :: String -> String
responseFor xs
  | null trimmed = "Fine. Be that way!"
  | isAllCaps && isQuestion = "Calm down, I know what I'm doing!"
  | isAllCaps = "Whoa, chill out!"
  | isQuestion = "Sure."
  | otherwise = "Whatever."
  where
    trimmed = allTrim xs
    allTrim = trimLeft . trimRight
    trimLeft = dropWhile C.isSpace
    trimRight = dropWhileEnd C.isSpace

    isQuestion = last trimmed == '?'
    isAllCaps = all C.isUpper onlyLetters && (not . null) onlyLetters
      where
        onlyLetters = filter C.isAlpha trimmed
